app/controllers/authentications_controller.rb

```diff
@@ -1,23 +1,63 @@
 class AuthenticationsController < Devise::OmniauthCallbacksController
+
   def schoology
     generic_oauth
   end
 
+  def google
+    generic_oauth
+  end
+
   private
 
   def generic_oauth
-    omniauth = request.env["omniauth.auth"]
+
+    omniauth    = request.env["omniauth.auth"]
+    origin      = request.env["omniauth.origin"]
+
     if extra = omniauth.extra
-      session[:portal_username] = extra.username
-      session[:portal_user_id]  = extra.user_id
-      session[:portal_domain]   = extra.domain
+
+      if extra.username.present?
+
+        #
+        # Handle Schoology oauth data.
+        #
+        session[:portal_username] = extra.username
+        session[:portal_user_id]  = extra.user_id
+        session[:portal_domain]   = extra.domain
+
+      else
+
+        #
+        # Handle google oauth data
+        #
+        parts = omniauth.info.email.split('@')
+        session[:portal_username] = parts[0]
+        session[:portal_user_id]  = omniauth.info.uid
+        session[:portal_domain]   = parts[1]            # What is domain?
+                                                        # In this case it
+                                                        # effectively stores 
+                                                        # the email....?
+      end
+
+      #
+      # For teachers we should set a valid email after creating
+      # the teacher user.
+      #
+      session[:omniauth_email]  = omniauth.info.email
+
+      #
+      # Return user to omniauth_origin after signup is complete.
+      #
+      session[:omniauth_origin] = origin
```

Comment:
I am not aware that this worked with schoology. 
This is an attempt to support "Log in from collection pages" from PT# 148097177
The intention is user is redirected back to the collections page they were viewing when they signed up.